src/session/content/message_row/photo.rs

```diff
@@ -0,0 +1,160 @@
+            let message = message.downcast_ref::<Message>().unwrap();
+
+            // Setup caption expression
+            let content_expression = gtk::PropertyExpression::new(
+                Message::static_type(),
+                gtk::NONE_EXPRESSION,
+                "content",
+            );
+            let text_expression = gtk::ClosureExpression::new(
+                |args| {
+                    let content = args[1].get::<BoxedMessageContent>().unwrap();
+                    if let MessageContent::MessagePhoto(data) = content.0 {
+                        parse_formatted_text(data.caption)
+                    } else {
+                        unreachable!();
+                    }
+                },
+                &[content_expression.upcast()],
+            );
+            let text_binding = text_expression.bind(&*self_.media, "caption", Some(message));
+            self_.binding.replace(Some(text_binding));
+
+            // Load photo
+            let handler_id =
+                message.connect_content_notify(clone!(@weak self as obj => move |message, _| {
+                    obj.update_photo(message);
+                }));
+            self_.handler_id.replace(Some(handler_id));
+            self.update_photo(message);
+        }
+
+        self_.old_message.set(self.message().as_ref());
+    }
+
+    fn update_photo(&self, message: &Message) {
+        if let MessageContent::MessagePhoto(data) = message.content().0 {
+            if let Some(photo_size) = data.photo.sizes.last() {
+                let self_ = imp::MessagePhoto::from_instance(self);
+
+                // Reset media widget
+                self_.media.set_paintable(None);
+                self_
+                    .media
+                    .set_aspect_ratio(photo_size.width as f64 / photo_size.height as f64);
+
+                if photo_size.photo.local.is_downloading_completed {
+                    self_.media.set_download_progress(1.0);
+                    self.load_photo_from_path(&photo_size.photo.local.path);
+                } else {
+                    self_.media.set_download_progress(0.0);
+                    self.download_photo(photo_size.photo.id, &message.chat().session());
+                }
+            }
+        }
+    }
+
+    fn download_photo(&self, file_id: i32, session: &Session) {
+        let (sender, receiver) = glib::MainContext::sync_channel::<File>(Default::default(), 5);
+
+        receiver.attach(
+            None,
+            clone!(@weak self as obj => @default-return glib::Continue(false), move |file| {
+                let self_ = imp::MessagePhoto::from_instance(&obj);
+
+                if file.local.is_downloading_completed {
+                    self_.media.set_download_progress(1.0);
+                    obj.load_photo_from_path(&file.local.path);
+                } else {
+                    let progress = file.local.downloaded_size as f64 / file.expected_size as f64 ;
```

Comment:
```suggestion
                    let progress = file.local.downloaded_size as f64 / file.expected_size as f64;
```

Reply:
I really have that rustfmt doesn't work inside macros...

Reply:
Yeah, I am with you. It's painful.